### DISCUSSION TITLE
feature: prometheus histogram metrics

### DISCUSSION THREAD

**Discussion by sbilello** [Netflix/dgs-framework#345] [Ideas]
## Describe the Feature Request
Add Histogram metrics support https://prometheus.io/docs/practices/histograms/
## Describe Preferred Solution
Annotation-based approach.

**Comment by berngp** [comment]
Hi @sbilello, we are using [Micrometer](https://micrometer.io/) as the mechanism to capture the metrics and they meters can be configured to [export percentiles](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#per-meter-properties). Its a matter of just setting `management.metrics.distribution.percentiles-histogram.gql=true`.

**Comment by berngp** [comment]
@sbilello I think unless there is a specific feature that you suggest for the DGS framework, this looks more like a Q&A discussion.

**Comment by sbilello** [comment]
This is cool! Thanks, @berngp! 
Yes, we can move this as Q&A but I was wondering if it would make sense to specify the percentiles as part of the configuration to reduce the number of buckets. See something like https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/annotation/Timed.java#L54

 I was also wondering why we have different labels for the metrics:

```
gql_query_seconds_bucket{gql_queryComplexity="25",outcome="success",le="0.001",} 0.0
gql_resolver_seconds_bucket{gql_field="Query.getUserByUserIdInput",outcome="success",le="7.158278826",} 20.0
```

I will work on a custom tag as described in the [doc](https://netflix.github.io/dgs/advanced/instrumentation/#further-tag-customization) unless we can not make consistent the tags between these 2 metrics.
 
I would like to show the duration in percentiles by `query name` for the `gql_resolver*` and the `gql_query*` metrics

**Comment by berngp** [comment]
Not sure I understand which labels are you referring to on `gql.query` vs `gql.resolver`.
The `gql.query` captures the entire scope of the GraphQL Query. Since each GraphQL Query can have multiple fields we can't capture them as part of the `gql.query` tag, I don't believe there is a way to express an array of values associated to a tag puls the cardinality will be problematic. The `gql.resolver` on the other hand can capture the _GraphQL Field_ and therefore is able to include it.

**Reply by sbilello** [reply]
I would like a `query_name` tag to be able to get the latency by `query_name` so I could be able to slice the latency per `query_name` across resolvers and data fetcher :)

**Reply by berngp** [reply]
In general, I think is an interesting idea. In order for us to support it out of the box we will have to control the cardinality of a normalized interpretation of the query name. We use Micrometer, and I am not aware there is something that will help us control a _tag's cardinality_ in Micrometer (for example, in Spectator there are [CardinalityLimiters](https://github.com/Netflix/spectator/blob/master/spectator-api/src/main/java/com/netflix/spectator/api/patterns/CardinalityLimiters.java)).

**In the meantime** you could implement a `DgsExecutionTagCustomizer`, the tags you define with it will get applied to the `gql.query` timer metric.

**Reply by sbilello** [reply]
![TagsForQueriesinUI](https://user-images.githubusercontent.com/1815376/119565167-7ad09e80-bd5e-11eb-8246-882757b17e77.png)
So a proposal would be to have 2 tags:
- tag for the query root name
- tag for hash of the payload
It looks to be a hash prefix the one shown in their UI

Other question:
Is there any interest to limit the Percentiles collected in the histogram with some extra `percentiles` array?
https://github.com/micrometer-metrics/micrometer/blob/main/micrometer-core/src/main/java/io/micrometer/core/annotation/Timed.java#L54

**Reply by berngp** [reply]
I'm working on adding additional tags that will do what you mentioned. Regarding the percentiles, the idea was to have developers limit that via configuration instead of fixing it via de annotation.

**Reply by sbilello** [reply]
Thanks, @berngp. Configuration => `Java Bean` or `Annotation` or `application.yml` ?

**Reply by berngp** [reply]
We are talking about two different things.

1. **The Percentiles in Micrometer.** Configuring the percentiles in micrometer is better done via its supported configuration instead of having the annotation express that explicitly. Ref [Micrometer's Adding percentiles, histograms, and SLA boundaries documentation](https://micrometer.io/docs/ref/spring/1.5#_adding_percentiles_histograms_and_sla_boundaries)

2. **Adding a GQL Query identifier to the tags.** This is a work in progress.

**Reply by sbilello** [reply]
Thanks for clarifying :)
